cheatsheets/Authentication_Cheat_Sheet.md

```diff
@@ -126,32 +126,32 @@ Example using pseudo-code for a login feature:
+    IS_VALID=LOOKUP_CREDENTIALS_IN_STORE(username, password_hash)
     IF NOT IS_VALID THEN
-        RETURN Error("Invalid User or Password!")    
+        RETURN Error("Invalid Username or Password!")    
     ENDIF
 ELSE
-   RETURN Error("Invalid User or Password!")
+   RETURN Error("Invalid Username or Password!")
 ENDIF
 ```
 
-With this implementation: We can have a time-based discrepancy factor because, if the user exists then two time-taking operations are performed otherwise only one if the user does not exist.
+It can be clearly seen that if the user doesn't exist, the application will directly throw out an error. Otherwise, when the user exists and the password doesn't, it is apparent that there will be more processing before the application errors out. In return, the response time will be different for the same error, allowing for the attacker to differentiate between a wrong username and a wrong password.
 
-*Second implementation do not using the "quick exit" approach:*
+*Second implementation without relying on the "quick exit" approach:*
 
 ```
 password_hash=HASH(password)
-IS_VALID=LOOKUP_CREDENTIALS_IN_STORE(login, password_hash)
+IS_VALID=LOOKUP_CREDENTIALS_IN_STORE(username, password_hash)
 IF NOT IS_VALID THEN
-   RETURN Error("Invalid User or Password!")
+   RETURN Error("Invalid Username or Password!")
 ENDIF
 ```
 
-With this implementation: The presence of a time-based discrepancy factor is less possible because, whatever if the user exists or not then two time-taking operations are always performed.
+This code will go through the same process no matter what the user or the password is, allowing the application to return in approximately the same response time.
 
-The main issue with the fact to return a generic error message is that it will decrease the *user experience* of the feature and can confuse a legitimate user because the generic error message will not allow him to understand the source of the issue preventing him to access to his account/data. The decision to return or not a *generic error message* can be determined according to the criticity of the application and the data managed. For example, for critical applications, we can decide that in case of trouble, a user will be redirected to the helpdesk and a *generic error message* will always be returned.
+The main issue with the fact to return a generic error message is that it will decrease the *user experience* of the feature and can confuse a legitimate user because the generic error message will not allow them to understand the source of the issue preventing them access to their account. The decision to return or not a *generic error message* can be determined according to the criticality of the application and the data managed. For example, for critical applications, team can decide that in case of trouble, a user will be redirected to the helpdesk and a *generic error message* will always be returned.
```

Comment:
What do you think about this rewording of this paragraph?
```
The problem with returning a generic error message for the user is a User Experience (UX) matter. A legitimate user might feel confused with the generic messages, thus making it hard for them to use the application, and might after several retries, leave the application because of its complexity. The decision to return a *generic error message* can be determined based on the criticality of the application and its data. For example, for critical applications, the team can decide that under the failure scenario, a user will always be redirected to the support page and a *generic error message* will be returned.
```

Reply:
Fix applied in coming commit.